fuse_optimizers/include/fuse_optimizers/batch_optimizer.h

```diff
@@ -108,9 +109,10 @@ class BatchOptimizer : public Optimizer
    * @param[in] private_node_handle A node handle in the node's private namespace
    */
   BatchOptimizer(
-    fuse_core::Graph::UniquePtr graph,
-    const ros::NodeHandle& node_handle = ros::NodeHandle(),
-    const ros::NodeHandle& private_node_handle = ros::NodeHandle("~"));
+    rclcpp::NodeOptions options,
+    std::string node_name = "batch_optimizer_node",
+    fuse_core::Graph::UniquePtr graph = nullptr
```

Comment:
I needed to be able to defer the initialization of the graph because a Node is required to pull the graph parameters to start the graph up.

The way the classes have been architected, one easy place to get access to a node is the optimizer object itself (but the graph can't be created while said Node is getting constructed.)

So later on, in the constructor body, the node is used if the graph isn't populated.

```
  if (!graph) {
    fuse_graphs::HashGraphParams hash_graph_params;
    hash_graph_params.loadFromROS(*this);
    graph_ = std::move(fuse_graphs::HashGraph::make_unique(hash_graph_params));
  }
```

Reply:
> I needed to be able to defer the initialization of the graph because a Node is required to pull the graph parameters to start the graph up.

Ah that's very interesting. The `Node` needs to exist before a `Graph` (or a `HashGraph` instance) can exist, but the node is being created inside the Optimizer, and the Optimizer has an argument that expects the `Graph` to already exist. Seems like a chicken and egg problem.

There's another issue: Optimizer doesn't allow lifecycle nodes to be used. A user would have to create a separate `LifecycleOptimizer` to create a lifecycle node inside.

Looking at the ROS 1 code, I notice the Optimizers are passed in node handles, which seems sort of like passing in nodes. What if we make the `Optimizer` accept `rclcpp::NodeInterfaces` instead? Then users can both create lifecycle nodes, and could initialize a Graph ahead of time if they wanted to. That would mean changing the constructor args here to `BatchOptimizer(fuse_core::Graph::UniquePtr graph, rclcpp::NodeInterfaces interfaces)`. The downside is now the user is forced to provide the interfaces argument instead of having a node created for them.

Reply:
I'll do: `BatchOptimizer(rclcpp::NodeInterfaces interfaces, fuse_core::Graph::UniquePtr graph)`

So that the graph is an optional param for potential deferred creation

Reply:
Fixed! https://github.com/locusrobotics/fuse/pull/307/commits/864acc55cadf39b4b23eea1b087b2169c6c059e9

Reply:
ping @svwilliams  This PR conveniently creates a `HashGraph` instance automatically if the given graph argument is `nullptr`. Previously the user was required to construct a graph and pass it in. Thoughts?